File: src/main/java/ee/leola/kassa/controllers/rest/RestBaseController.java
package ee.leola.kassa.controllers.rest;

import java.io.Serializable;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.BeanUtils;
import org.springframework.data.repository.CrudRepository;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@Slf4j
public abstract class RestBaseController<T, ID extends Serializable> {

  protected CrudRepository<T, ID> repo;

  public RestBaseController(CrudRepository<T, ID> repo) {
    this.repo = repo;
  }

  @RequestMapping
  @ResponseBody
  public Iterable<T> listAll() {
    return repo.findAll();
  }

  @RequestMapping(
      method = RequestMethod.POST,
      consumes = {MediaType.APPLICATION_JSON_VALUE})
  @ResponseBody
  public T create(@RequestBody T json) {
    log.info("create() with body {} of type {}", json, json.getClass());
    return repo.save(json);
  }

  @RequestMapping(value = "/{id}", method = RequestMethod.GET)
  @ResponseBody
  public T get(@PathVariable ID id) {
    return repo.findById(id).orElseThrow(IllegalStateException::new);
  }

  @RequestMapping(
      value = "/{id}",
      method = RequestMethod.POST,
      consumes = {MediaType.APPLICATION_JSON_VALUE})
  @ResponseBody
  public ResponseEntity<T> update(@PathVariable ID id, @RequestBody T json) {
    log.info("update() of id {} with body {}", id, json);
    log.info("T json is of type {}", json.getClass());

    T entity = repo.findById(id).orElseThrow(IllegalStateException::new);
    try {
      BeanUtils.copyProperties(json, entity);
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
    T updated = repo.save(entity);
    log.info("updated entity: {}", updated);
    return new ResponseEntity<>(updated, HttpStatus.OK);
  }

  @RequestMapping(value = "/{id}", method = RequestMethod.DELETE)
  @ResponseBody
  public void delete(@PathVariable ID id) {
    repo.deleteById(id);
  }
}

File: src/main/kotlin/ee/leola/kassa/controllers/rest/RestBaseController.kt
package ee.leola.kassa.controllers.rest

import org.slf4j.LoggerFactory
import org.springframework.beans.BeanUtils
import org.springframework.data.repository.CrudRepository
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.io.Serializable

abstract class RestBaseController<T : Any, ID : Serializable>(protected val repo: CrudRepository<T, ID>) {

    private val logger = LoggerFactory.getLogger(this::class.java)

    @RequestMapping
    @ResponseBody
    fun listAll(): Iterable<T> = repo.findAll()

    @ResponseBody
    @PostMapping(consumes = [MediaType.APPLICATION_JSON_VALUE])
    open fun create(@RequestBody json: T): T = repo.save(json).also {
        logger.info("create() with body {} of type {}", json, json::class)
    }

    @ResponseBody
    @GetMapping("/{id}")
    fun get(@PathVariable id: ID): T = repo.findById(id).orElseThrow()

    @ResponseBody
    @PostMapping("/{id}", consumes = [MediaType.APPLICATION_JSON_VALUE])
    open fun update(@PathVariable id: ID, @RequestBody json: T): ResponseEntity<T> {
        logger.info("update() of id {} with body {}", id, json)
        logger.info("T json is of type {}", json::class)
        val entity = repo.findById(id).orElseThrow()
        BeanUtils.copyProperties(json, entity)
        val updated = repo.save(entity)
        logger.info("updated entity: {}", updated)
        return ResponseEntity(updated, HttpStatus.OK)
    }

    @ResponseBody
    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: ID) = repo.deleteById(id)
}